jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -2368,21 +2377,58 @@ private ClassDefiner classDefiner(byte[] bytes, ClassOption... classOptions) thr
 				throw new IllegalAccessException();
 			}
 
+			int flags = 0;
+			for (ClassOption opt : classOptions) {
+				flags |= opt.toFlag();
+			}
+			MethodHandleNatives.checkClassBytes(bytes);
+			String targetClassName = getClassName(bytes);
+			return makeHiddenClassDefiner(targetClassName, bytes, flags);
+			
+		}
+
+		private String getClassName(byte[] bytes) {
+			String targetClassName;
 			ClassReader cr;
 			try {
 				cr = new ClassReader(bytes);
+
+				int thisClassIndex = cr.readUnsignedShort(cr.header + 2);
+				char[] buffer = new char[cr.getMaxStringLength()];
+				Object thisClass = cr.readConst(thisClassIndex, buffer);
+				if (!(thisClass instanceof Type)) {
+					throw new ClassFormatError();
+				}
+				Type type = (Type)thisClass;
+				if (!type.getDescriptor().startsWith("L")) {
+					throw new ClassFormatError();
+				}
+				targetClassName = type.getClassName();
```

Comment:
All this code can go away if the native returns the class name string

Reply:
I have a local change that address all the comments here. i.e. makes MethodHandleNatives.checkClassBytes() returns the class name and with the code above removed. But I am failing some hidden class tests:

1. There are test cases that make thisClass not an instance of Type and the descriptor does not start with "L", these tests asserts ClassFormatError to be thrown. These checks are not there inside j9bcutil_readClassFileBytes(). So I guess we cannot remove these checks. 

```
if (!(thisClass instanceof Type)) {
	throw new ClassFormatError();
}
if (!type.getDescriptor().startsWith("L")) {
	throw new ClassFormatError();
}
```
2. After adding the above checks back, I am still failing one test. 
Because I am getting the class name from j9bcutil_readClassFileBytes(), it has to return after constant pool are read and  classfile->thisClass is set. But before setting classfile->thisClass, j9bcutil_readClassFileBytes()
 checks on CFR_ACC_MODULE and throw ClassFormatError on CFR_ACC_MODULE. Hidden class tests expects an IllegalArgumentException if CFR_ACC_MODULE is set.

Reply:
> There are test cases that make thisClass not an instance of Type and the descriptor does not start with "L", these tests asserts ClassFormatError to be thrown. 

Can you provide examples of the invalid `thisClass` values?  I can understand the "L" case but need clarification on the `instanceof Type` case and why it wouldn't be caught by the cfreader code (unless maybe we don't check it till staticverify runs?)